android/src/main/java/com/mattermost/networkclient/APIClientModule.kt

```diff
@@ -4,99 +4,124 @@ import com.facebook.react.bridge.*
 import okhttp3.OkHttpClient
 import okhttp3.Request
 import okhttp3.RequestBody.Companion.toRequestBody
-import com.facebook.react.bridge.Arguments;
+import okio.IOException
 
 class APIClientModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
 
-    var client = OkHttpClient();
+    var sessionsClient = mutableMapOf<String, OkHttpClient.Builder>();
     var sessions = mutableMapOf<String, Request.Builder>()
 
     override fun getName(): String {
         return "APIClient"
     }
 
     @ReactMethod
-    fun createClientFor(baseUrl: String, config: ReadableMap, promise: Promise){
+    fun createClientFor(baseUrl: String, options: ReadableMap, promise: Promise) {
         try {
+            // Create the client and request builder
+            sessionsClient[baseUrl] = OkHttpClient().newBuilder();
             sessions[baseUrl] = Request.Builder().url(baseUrl);
-            promise.resolve(sessions[baseUrl]);
-        } catch (err: Throwable){
+
+            // Attach client options
+            sessionsClient[baseUrl]?.parseOptions(options);
+
+            // Return stringified client for success
+            promise.resolve(Unit)
+        } catch (err: Throwable) {
             promise.reject(err)
         }
     }
 
     @ReactMethod
-    fun getClientHeadersFor(baseUrl: String, promise: Promise){
+    fun getClientHeadersFor(baseUrl: String, promise: Promise) {
         try {
-            val headersMap = Arguments.createMap()
-            sessions[baseUrl]!!.build().headers.forEach{(k,v) -> headersMap.putString(k, v)}
-            promise.resolve(headersMap)
+            promise.resolve(sessions[baseUrl]?.build()?.headers?.readableMap())
         } catch (error: Error) {
             promise.reject(error)
         }
     }
 
     @ReactMethod
-    fun addClientHeadersFor(baseUrl: String, headers: ReadableMap, promise: Promise){
-        for ((k,v) in headers.toHashMap()){
-            sessions[baseUrl]!!.addHeader(k,v as String);
+    fun addClientHeadersFor(baseUrl: String, headers: ReadableMap, promise: Promise) {
+        try {
+            sessions[baseUrl]?.addReadableMap(headers)
+            promise.resolve(Unit);
+        } catch (error: Error) {
+            promise.reject(error)
```

Comment:
I think we should probably make a set of error cases that are handled the same way on both platforms? Happy to write up an error interface in our `@types` that the JS side can expect and we can implement on the native side; though I think it might be bigger task for outside this PR...

@enahum / @avinashlng1080 - any thoughts?

Reply:
I think we should define a map so that we have consistent errors for both platforms instead of returning the raw error. I would expect it to be raw just for the responses

Reply:
Yes a map will be super helpful.  I'm thinking of something like the following can be standardized across both platforms and it may also become useful in the future when we'll implement the 'logger feature'.

```
{
  errorCode : 'xxxxx', // The error code  - e.g. Network_Header_xxxx
  errorClassMethod: 'xxxx',  // APIClientModule - addClientHeadersFor
  errorDescription: 'xxxx xxxxxx xxxxx', // A text description of the error
  errorObject: { xxxx } // The full error object as generated.
}
```